docs/docsite/rst/dev_guide/developing_modules_general.rst

```diff
@@ -1,333 +1,313 @@
+-  `Local/direct module testing <#localdirect-module-testing>`__
+-  `Playbook module testing <#playbook-module-testing>`__
+-  `Debugging (local) <#debugging-local>`__
+-  `Debugging (remote) <#debugging-remote>`__
+-  `Unit testing <#unit-testing>`__
+-  Integration testing (coming soon)
+-  `Communication and development
+   support <#communication-and-development-support>`__
+-  `Credit <#credit>`__
 
-Reading the modules that come with Ansible (linked above) is a great way to learn how to write
-modules.   Keep in mind, though, that some modules in Ansible's source tree are internalisms,
-so look at :ref:`service` or :ref:`yum`, and don't stare too close into things like ``async_wrapper`` or
-you'll turn to stone.  Nobody ever executes ``async_wrapper`` directly.
 
-Ok, let's get going with an example.  We're going to use Python.  For starters, save this as a file named :file:`timetest.py`
+Environment setup
+=================
 
-.. code-block:: python
+1. Clone the Ansible repository:
+   ``$ git clone https://github.com/ansible/ansible.git``
+2. Change directory into the repository root dir: ``$ cd ansible``
+3. Create a virtual environment: ``$ python3 -m venv venv`` (or for
+   Python 2 ``$ virtualenv venv``. Note, this requires you to install
+   the virtualenv package: ``$ pip install virtualenv``)
+4. Activate the virtual environment: ``$ . venv/bin/activate``
+5. Install development requirements:
+   ``$ pip install -r requirements.txt``
+6. Run the environment setup script for each new dev shell process:
+   ``$ . hacking/env-setup``
 
-    #!/usr/bin/python
-
-    import datetime
-    import json
-
-    date = str(datetime.datetime.now())
-    print(json.dumps({
-        "time" : date
-    }))
-
-.. _module_testing:
-
-Testing Your Module
-```````````````````
-
-There's a useful test script in the source checkout for Ansible:
-
-.. code-block:: shell-session
+.. note:: After the initial setup above, every time you are ready to start
+   developing Ansible you should be able to just run the following from the
+   root of the Ansible repo:
+   ``$ . venv/bin/activate && . hacking/env-setup``
 
-    git clone git://github.com/ansible/ansible.git
-    source ansible/hacking/env-setup
 
-For instructions on setting up Ansible from source, please see
-:doc:`../intro_installation`.
+New module development
+======================
 
-Let's run the script you just wrote with that:
+If you are creating a new module that doesn't exist, you would start
+working on a whole new file. Here is an example:
 
-.. code-block:: shell-session
+-  Navigate to the directory that you want to develop your new module
+   in. E.g. ``$ cd lib/ansible/modules/cloud/azure/``
+-  Create your new module file: ``$ touch my_new_test_module.py``
+-  Paste this simple into the new module file: (explanation in comments)::
 
-    ansible/hacking/test-module -m ./timetest.py
-
-You should see output that looks something like this:
+    #!/usr/bin/python
```

Comment:
I think we should add a more realworld example in,

Also I think whatever we add should pass 
`ansible-test sanity --python 2.7 --test validate-modules  lib/ansible/modules/commands/foobar.py`

Currently
```
ERROR: lib/ansible/modules/commands/foobar.py:0:0: E105 GPLv3 license header not found
ERROR: lib/ansible/modules/commands/foobar.py:0:0: E319 RETURN.message.returned: required key not provided @ data['returned']. Got None
ERROR: lib/ansible/modules/commands/foobar.py:0:0: E319 RETURN.message.type: required key not provided @ data['type']. Got None
ERROR: lib/ansible/modules/commands/foobar.py:0:0: E319 RETURN.original_message.returned: required key not provided @ data['returned']. Got None
ERROR: lib/ansible/modules/commands/foobar.py:33:1: E302 DOCUMENTATION is not valid YAML
```